src/Microsoft.AspNetCore.Mvc.Core/Internal/RouteTemplateWriter.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Routing.Template;
+
+namespace Microsoft.AspNetCore.Mvc.Internal
+{
+    internal static class RouteTemplateWriter
+    {
+        public static string ToString(IEnumerable<TemplateSegment> routeSegments)
+        {
+            return string.Join("/", routeSegments.Select(s => ToString(s)));
+        }
+
+        private static string ToString(TemplateSegment templateSegment)
+        {
+            return string.Join(string.Empty, templateSegment.Parts.Select(p => ToString(p)));
+        }
+
+        private static string ToString(TemplatePart templatePart)
+        {
+            if (templatePart.IsParameter)
+            {
+                var partText = "{";
+                if (templatePart.IsCatchAll)
+                {
+                    partText += "*";
+                }
+                partText += templatePart.Name;
+                foreach (var item in templatePart.InlineConstraints)
+                {
+                    partText += ":";
+                    partText += item.Constraint;
+                }
+                if (templatePart.DefaultValue != null)
+                {
+                    partText += "=";
+                    partText += templatePart.DefaultValue;
+                }
+                if (templatePart.IsOptional)
+                {
+                    partText += "?";
+                }
+                partText += "}";
+
+                return partText;
+            }
+            else
+            {
+                return templatePart.Text;
+            }
+        }
```

Comment:
This is dope. Worth putting this in routing?

Reply:
There is a cut down version in routing for debugging - https://github.com/aspnet/Routing/blob/93d20ec78c49b9adf190c76fe7b3d7e8036bc7b9/src/Microsoft.AspNetCore.Routing/Template/TemplatePart.cs#L55-L65

Could make it a real `ToString` on RouteTemplate\RouteSegment\RoutePart. Or a `WriteTo(TextWriter textWriter)`?

I don't know how people use routing and whether it is worth it 🤷‍♂️

Reply:
I don't feel strongly about it either way. If it's low cost it might make sense to move these to routing as ToString's

Reply:
I'll keep it inside MVC for now. I would like to have the templates built up via a StringBuilder rather than many concats and `string.Join`s to keep allocations down.